packages/access-client/src/agent.js

```diff
@@ -175,21 +176,24 @@ export class Agent {
     const values = []
     for (const [, value] of this.#data.delegations) {
       // check expiration
-      if (
-        !isExpired(value.delegation) && // check if delegation can be used
-        !isTooEarly(value.delegation)
-      ) {
-        // check if we need to filter for caps
-        if (Array.isArray(caps) && caps.length > 0) {
-          for (const cap of _caps) {
-            if (canDelegateCapability(value.delegation, cap)) {
-              _caps.delete(cap)
-              values.push(value)
-            }
+      if (isExpired(value.delegation)) {
+        continue
+      }
+      if (isTooEarly(value.delegation)) {
+        continue
+      }
+
+      if (Array.isArray(caps) && caps.length > 0) {
+        // caps param is provided. Ensure that the delegations we're looping over can delegate to these caps
+        for (const cap of _caps) {
+          if (canDelegateCapability(value.delegation, cap)) {
+            values.push(value)
+            _caps.delete(cap)
```

Comment:
Can we make it so that this includes all delegations that can delegate this capability, not just the first?

From your investigation, this seemed to be the issue right?

Reply:
First, fyi I just now undid my changes to this `#delegations()` fn and reverted it back to how it was. While I was debugging this yesterday and trying to recreate as a test case and then try ways of making it pass, I had originally tried adding some smarts to this function to make it `invocationAudience` and session proof aware. But then I realized that would be doing some work twice given that the whole way the `proofs()` method works is calling `this.#delegations` but then also tacking on sessionProofs that are relevant to the result. So then I moved the invocation audience aware logic out of here and a layer higher into the `proofs()` method and its usage of `getSessionProofs`.

> Can we make it so that this includes all delegations that can delegate this capability, not just the first?

I'm not against that, but it wasn't required to fix the bug that I was seeing in my setup and in the test, and this is the logic that's been here for awhile, so I'm hesitant to change it if I don't have a test case that illustrates why.

> From your investigation, this seemed to be the issue right?

Yesterday I realized that in the situation I personally ran into in [here](https://github.com/web3-storage/w3cli/issues/113) and then I think captured in test case, that wasn't actually the underlying issue.

The underlying issue was more directly that in the caller of this function, `Agent#proofs`, even if the right proof was returned, it would then use `getSessionProofs` to add a relevant session proof to the invocation, but that could return a session proof that had an issuer that was unknown to the invocation target. So then when the invocation target (e.g. `did:web:staging.web3.storage`) got a session proof issued by some unknown id (e.g. `did:web:web3.storage`), it would lead to that [error](https://github.com/web3-storage/w3cli/issues/113)

So once I realized that was the root cause of the issue for w3cli 113, I could fix things not by touching this function and instead by making `proofs()` method have an optional `invocationTarget` param and then passing that to `getSessionProofs` so that getSessionProofs could return the appropriate session proof in this weird case where there are multiple session proof issuers (which is unusual for end-user, but not unusual for a w3infra dev)

Reply:
I ended up finding a test that failed unless I removed the `caps.delete()` so i removed it https://github.com/web3-storage/w3up/compare/cb90340...79a544dda9c382a73b7dd6383729c9b09b72834b#diff-d14594f7f9441ea59499c7e8f4d6ebb32ccb406de725d9c89cb3e53f1d63e37cR394